### PR TITLE
use poetry for publication

### PR DESCRIPTION
**Pull request by darthtrevino**
Signed-off-by: Chris Trevino <chtrevin@microsoft.com>

Fixes #

Changes proposed in this pull request:
- Use poetry 1.2 in main.yml
- Use poetry to install dynamic versioning plugin. This is the recommended method for 1.2.0+
- Use `poetry publish` in release task

## Before submitting

- [X] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/dodiscover/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.

## After submitting

- [X] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -91,8 +91,8 @@ graph_func = ['pywhy-graphs']
 viz = ['pygraphviz']
 
 [build-system]
-requires = ["poetry-core>=1.0.0"]
-build-backend = "poetry.core.masonry.api"
```
Comment (adam2392) [inline pyproject.toml:95]: What's the motivation for this change?

**Review by adam2392 — APPROVED** [review]
LGTM!

**Review by adam2392 — CHANGES_REQUESTED** [review]
Seems CI is running into some errors.

File: pyproject.toml (hunk 1 above)
Comment (darthtrevino) [reply pyproject.toml:95]: According to [their docs](https://github.com/mtkennerly/poetry-dynamic-versioning), this is the recommended configuration for the dynamic-versioning plugin in 1.2.0

Comment (darthtrevino) [reply pyproject.toml:95]: It looks like swapping the build-backend caused the sdist checker to fail. Reverting to pip-based install.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -145,7 +150,7 @@ jobs:
       #   run: |
       #     pip uninstall -yq networkx
       #     pip install --progress-bar off git+https://github.com/networkx/networkx
-      - name: Run pytest  # headless via Xvfb on linux
+      - name: Run pytest # headless via Xvfb on linux
```
Comment (adam2392) [inline .github/workflows/main.yml:153]:
```suggestion
      - name: Run pytest
```

File: doc/api.rst [hunk 3]
```diff
@@ -37,8 +37,8 @@ Graph protocols
 .. autosummary::
    :toctree: generated/
 
-   GraphProtocol
-   EquivalenceClassProtocol
+   Graph
+   EquivalenceClass
```
Comment (adam2392) [inline doc/api.rst:41]: I think the DCO rebasing makes these PR diffs nonsensical for some reason. E.g. this diff isn't technically part of this PR, but the previous PR... rip. idk how to fix (usually it is to rebase, but seeing as how you did..)

**Review by adam2392 — CHANGES_REQUESTED** [review]
Hey @darthtrevino since you're a more experienced dev, can we try to fix the diff since it seems to show the diff from #35 too for some reason...

- Perhaps `git pull main`
- `git rebase ci/poetry_publish main` ?

If it doesn't work... then I guess we'll just have to deal w/ the DCO weirdness that rebases things weirdly?

The code itself LGTM. The build GH action seems like something that we'll have to test when we slate the first release, where someone needs to upload their PYPI admin username/pw to the GH secrets. But that can be saved for a later date... Easy enough fix.

**Review by adam2392 — APPROVED** [review]
I think the diff thing, we can handle in #42...
